File: swift/CubingCore/Sources/CubingCore/Openings.swift
/// Cross-first and joint cross/pair routes to one or two solved pairs.
public enum Openings {
    public struct Step: Sendable {
        public let label: String
        public let moves: [Move]
    }
    public struct Opening: Sendable {
        public let moves: [Move]
        public let steps: [Step]
        public let solvedSlots: [String]
    }
    public struct Result: Sendable {
        public let crossPlusOne: [Opening]
        public let crossPlusTwo: [Opening]
    }
    private struct Seed {
        let state: CubeState
        let moves: [Move]
        let steps: [Step]
    }

    private static func combineTurns(_ moves: [Move]) -> [Move] {
        var combined: [Move] = []
        for move in moves {
            guard let previous = combined.last, previous.familyIndex == move.familyIndex else {
                combined.append(move)
                continue
            }
            combined.removeLast()
            let amount = MoveTables.normalizeAmount(previous.amount + move.amount)
            if amount != 0 { combined.append(Move(familyIndex: move.familyIndex, amount: amount)) }
        }
        return combined
    }

    public static func search(
        _ raw: CubeState, _ crossFace: Int, xcrosses: [[Move]]
    ) -> Result {
        let state = raw.normalized
        let slots = geometry[crossFace].slots
        func solved(_ state: CubeState) -> [Slot] { slots.filter { isSlotSolved(state, $0) } }
        var plusOne: [Opening] = []
        var plusTwo: [Opening] = []

        func collect(_ moves: [Move], _ steps: [Step], _ after: CubeState) {
            let built = solved(after)
            let solvedSlots = built.map(\.name)
            let combined = combineTurns(moves)
            let opening: Opening
            if combined.count < moves.count {
                opening = Opening(
                    moves: combined,
                    steps: [Step(label: "cross + \(built.map(Colours.slot).joined(separator: " + "))",
                                 moves: combined)],
                    solvedSlots: solvedSlots)
            } else {
                opening = Opening(moves: moves, steps: steps, solvedSlots: solvedSlots)
            }
            if solvedSlots.count >= 1 { plusOne.append(opening) }
            if solvedSlots.count >= 2 { plusTwo.append(opening) }
        }

        var crossOptions = SearchOptions()
        crossOptions.maxExtra = 1
        crossOptions.maxSolutions = 24
        crossOptions.maxSolutionsPerDepth = 12
        crossOptions.maxNodes = 100_000
        let cross = enumerateCross(state, crossFace, crossOptions)
        var seeds: [CubeState: Seed] = [:]
        var seedOrder: [CubeState] = []
        for moves in cross.candidates.map(\.moves) + xcrosses {
            let after = state.applying(moves)
            let built = solved(after)
            let label = built.isEmpty ? "cross" : "cross + \(built.map(Colours.slot).joined(separator: " + "))"
            let steps = [Step(label: label, moves: moves)]
            if let previous = seeds[after] {
                if previous.moves.count > moves.count {
                    seeds[after] = Seed(state: after, moves: moves, steps: steps)
                }
            } else {
                seeds[after] = Seed(state: after, moves: moves, steps: steps)
                seedOrder.append(after)
            }
            collect(moves, steps, after)
        }
        func estimate(_ seed: Seed) -> Int {
            let needed = max(0, 2 - solved(seed.state).count)
            let distances = slots.filter { !isSlotSolved(seed.state, $0) }
                .map { pairDistance(seed.state, $0) }.sorted()
            return seed.moves.count + (needed == 0 ? 0 : distances[needed - 1])
        }
        func sortSeeds(_ values: [Seed]) -> [Seed] {
            values.enumerated().sorted { a, b in
                let left = estimate(a.element), right = estimate(b.element)
                return left == right ? a.offset < b.offset : left < right
            }.map(\.element)
        }
        let ordered = sortSeeds(seedOrder.map { seeds[$0]! })
        var groups: [String: [Seed]] = [:]
        var groupOrder: [String] = []
        for seed in ordered {
            let key = "\(seed.moves.count):\(solved(seed.state).map(\.name).joined(separator: ","))"
            if groups[key] == nil { groups[key] = []; groupOrder.append(key) }
            groups[key]!.append(seed)
        }
        var selected = Array(groupOrder.prefix(8).map { groups[$0]!.removeFirst() })
        let leftovers = sortSeeds(groupOrder.flatMap { groups[$0]! })
        for seed in leftovers {
            if selected.count >= 8 { break }
            selected.append(seed)
        }

        for seed in selected {
            var options = Lookahead.Options()
            options.maxNodes = 300_000
            let result = Lookahead.continueF2L(seed.state, crossFace, targetPairs: 2, options: options)
            for continuation in result.prefixes + result.candidates {
                var after = seed.state
                var moves = seed.moves
                var steps = seed.steps
                for step in continuation.steps {
                    after = after.applying(step.moves)
                    moves += step.moves
                    let slot = slots.first { $0.name == step.slot }!
                    steps.append(Step(label: Colours.slot(slot), moves: step.moves))
                    collect(moves, steps, after)
                }
            }
        }
        for first in 0..<slots.count {
            for second in (first + 1)..<slots.count {
                let pair = [slots[first], slots[second]]
                var options = insertionDefaults()
                options.maxDepth = 11
                options.maxSolutions = 3
                options.maxNodes = 50_000
                let result = enumerateF2LInsertion(
                    state, crossFace, pair[0], options: options, preserve: [pair[1]])
                for candidate in result.candidates {
                    collect(candidate.moves,
                            [Step(label: "cross + \(pair.map(Colours.slot).joined(separator: " + "))",
                                  moves: candidate.moves)],
                            state.applying(candidate.moves))
                }
            }
        }
        func unique(_ openings: [Opening]) -> [Opening] {
            var keyed: [[Move]: Opening] = [:]
            var order: [[Move]] = []
            for opening in openings {
                if keyed[opening.moves] == nil { order.append(opening.moves) }
                keyed[opening.moves] = opening
            }
            return order.map { keyed[$0]! }
        }
        return Result(crossPlusOne: unique(plusOne), crossPlusTwo: unique(plusTwo))
    }
}

File: packages/planner/src/openings.ts
/** Cross-first and joint cross/pair plans, compared at the same two-pair goal. */
import { applyMoves, makeMove, normalizeOrientation, type CubeState, type Face, type Move } from "@cubing-companion/engine";
import { GEOMETRY, isSlotSolved, slotName } from "@cubing-companion/analysis";
import { enumerateCross, enumerateF2LInsertion, pairDistance } from "@cubing-companion/solver";
import { continueF2L } from "./lookahead.ts";
import { slotColours } from "./colours.ts";

export interface OpeningStep {
  readonly label: string;
  readonly moves: readonly Move[];
}

export interface Opening {
  readonly moves: readonly Move[];
  readonly steps: readonly OpeningStep[];
  readonly solvedSlots: readonly string[];
}

interface Seed {
  readonly state: CubeState;
  readonly moves: readonly Move[];
  readonly steps: OpeningStep[];
}

/** Cancelling a boundary can remove an intermediate milestone, so label the result jointly. */
function combineTurns(moves: readonly Move[]): Move[] {
  const combined: Move[] = [];
  for (const move of moves) {
    const previous = combined[combined.length - 1];
    if (previous?.family !== move.family) combined.push(move);
    else {
      combined.pop();
      const merged = makeMove(move.family, previous.amount + move.amount);
      if (merged) combined.push(merged);
    }
  }
  return combined;
}

export function searchOpenings(
  raw: CubeState,
  crossFace: Face,
  xcrosses: readonly { readonly searchMoves: readonly Move[] }[],
): { crossPlusOne: Opening[]; crossPlusTwo: Opening[] } {
  const state = normalizeOrientation(raw);
  const slots = GEOMETRY[crossFace]!.slots;
  const solved = (position: CubeState) => slots.filter((slot) => isSlotSolved(position, slot));
  const crossPlusOne: Opening[] = [];
  const crossPlusTwo: Opening[] = [];
  const collect = (moves: readonly Move[], steps: readonly OpeningStep[], after: CubeState) => {
    const built = solved(after);
    const solvedSlots = built.map(slotName);
    const combined = combineTurns(moves);
    const opening = combined.length < moves.length ? {
      moves: combined,
      steps: [{ label: `cross + ${built.map(slotColours).join(" + ")}`, moves: combined }],
      solvedSlots,
    } : { moves, steps, solvedSlots };
    if (solvedSlots.length >= 1) crossPlusOne.push(opening);
    if (solvedSlots.length >= 2) crossPlusTwo.push(opening);
  };

  const cross = enumerateCross(state, crossFace, {
    maxExtra: 1,
    maxSolutions: 24,
    maxSolutionsPerDepth: 12,
    maxNodes: 100_000,
  });
  const seeds = new Map<string, Seed>();
  for (const moves of [...cross.candidates.map((c) => c.moves), ...xcrosses.map((c) => c.searchMoves)]) {
    const after = applyMoves(state, moves);
    const built = solved(after);
    const label = built.length === 0 ? "cross" : `cross + ${built.map(slotColours).join(" + ")}`;
    const previous = seeds.get(after.key());
    if (!previous || previous.moves.length > moves.length) {
      seeds.set(after.key(), { state: after, moves, steps: [{ label, moves }] });
    }
    collect(moves, [{ label, moves }], after);
  }
  const estimate = (seed: { state: CubeState; moves: readonly Move[] }) => {
    const needed = Math.max(0, 2 - solved(seed.state).length);
    const distances = slots.filter((slot) => !isSlotSolved(seed.state, slot))
      .map((slot) => pairDistance(seed.state, slot)).sort((a, b) => a - b);
    return seed.moves.length + (needed === 0 ? 0 : distances[needed - 1]!);
  };
  // Reserve seeds from each cross length and each built-slot set. A crowd of optimal crosses
  // must not consume every place before a +1 cross or an x-cross can be explored.
  const groups = new Map<string, Seed[]>();
  for (const seed of [...seeds.values()].sort((a, b) => estimate(a) - estimate(b))) {
    const key = `${seed.moves.length}:${solved(seed.state).map(slotName).join(",")}`;
    const group = groups.get(key) ?? [];
    group.push(seed);
    groups.set(key, group);
  }
  const selected = [...groups.values()].map((group) => group.shift()!).slice(0, 8);
  for (const seed of [...groups.values()].flat().sort((a, b) => estimate(a) - estimate(b))) {
    if (selected.length >= 8) break;
    selected.push(seed);
  }
  for (const seed of selected) {
    const result = continueF2L(seed.state, crossFace, 2, { maxNodes: 300_000 });
    for (const continuation of [...result.prefixes, ...result.candidates]) {
      let after = seed.state;
      let moves = [...seed.moves];
      const steps = [...seed.steps];
      for (const step of continuation.steps) {
        after = applyMoves(after, step.moves);
        moves = [...moves, ...step.moves];
        const slot = slots.find((s) => slotName(s) === step.slot)!;
        steps.push({ label: slotColours(slot), moves: step.moves });
        collect(moves, [...steps], after);
      }
    }
  }

  // Search joint two-pair goals too. This can find an XX-cross that never passes through any
  // retained cross or x-cross seed. Each of the six slot combinations receives an equal budget.
  for (let first = 0; first < slots.length; first++) {
    for (let second = first + 1; second < slots.length; second++) {
      const pair = [slots[first]!, slots[second]!] as const;
      const result = enumerateF2LInsertion(state, crossFace, pair[0], {
        preserve: [pair[1]],
        maxDepth: 11,
        maxSolutions: 3,
        maxNodes: 50_000,
      });
      for (const candidate of result.candidates) {
        collect(candidate.moves, [{ label: `cross + ${pair.map(slotColours).join(" + ")}`, moves: candidate.moves }],
          applyMoves(state, candidate.moves));
      }
    }
  }
  // A route can solve extra pairs incidentally. Count the actual result, and deduplicate identical
  // move sequences so a joint solution found for two target slots is shown only once.
  const unique = (openings: Opening[]) => [...new Map(openings.map((opening) => [
    opening.moves.map((move) => `${move.family}${move.amount}`).join(" "), opening,
  ])).values()];
  return { crossPlusOne: unique(crossPlusOne), crossPlusTwo: unique(crossPlusTwo) };
}
